catroid/src/org/catrobat/catroid/web/ServerCalls.java

```diff
@@ -82,102 +103,162 @@
 
 	public static boolean useTestUrl = false;
 	private String resultString;
-	private ConnectionWrapper connection;
 	private String emailForUiTests;
-	private int uploadStatusCode;
+	private final OkHttpClient okHttpClient;
+	private final Gson gson;
 
 	private ServerCalls() {
-		connection = new ConnectionWrapper();
+		okHttpClient = new OkHttpClient();
+		okHttpClient.setConnectionSpecs(Arrays.asList(ConnectionSpec.MODERN_TLS));
+		gson = new Gson();
 	}
 
 	public static ServerCalls getInstance() {
 		return INSTANCE;
 	}
 
-	// used for mock object testing
-	public void setConnectionToUse(ConnectionWrapper connection) {
-		this.connection = connection;
-	}
-
 	public void uploadProject(String projectName, String projectDescription, String zipFileString, String userEmail,
-			String language, String token, String username, ResultReceiver receiver, Integer notificationId,
-			Context context) throws WebconnectionException {
-		if (emailForUiTests != null) {
-			userEmail = emailForUiTests;
-		}
+							  String language, String token, String username, ResultReceiver receiver, Integer notificationId,
+							  Context context) throws WebconnectionException {
+
+		Preconditions.checkNotNull(context, "Context cannot be null!");
+
+		userEmail = emailForUiTests == null ? userEmail : emailForUiTests;
+		userEmail = userEmail == null ? "" : userEmail;
 
 		try {
 			String md5Checksum = Utils.md5Checksum(new File(zipFileString));
 
-			HashMap<String, String> postValues = new HashMap<String, String>();
-			postValues.put(PROJECT_NAME_TAG, projectName);
-			postValues.put(PROJECT_DESCRIPTION_TAG, projectDescription);
-			postValues.put(USER_EMAIL, userEmail == null ? "" : userEmail);
-			postValues.put(PROJECT_CHECKSUM_TAG, md5Checksum);
-			postValues.put(Constants.TOKEN, token);
-			postValues.put(Constants.USERNAME, username);
-
-			if (language != null) {
-				postValues.put(DEVICE_LANGUAGE, language);
+			final String serverUrl = useTestUrl ? TEST_FILE_UPLOAD_URL_HTTP : FILE_UPLOAD_URL;
+
+			Log.v(TAG, "Url to upload: " + serverUrl);
+
+			File file = new File(zipFileString);
+
+			RequestBody requestBody = new MultipartBuilder()
+					.type(MultipartBuilder.FORM)
+					.addFormDataPart(
+							FILE_UPLOAD_TAG,
+							ProjectUploadService.UPLOAD_FILE_NAME,
+							RequestBody.create(MEDIA_TYPE_ZIPFILE, file))
+					.addFormDataPart(
+							PROJECT_NAME_TAG,
+							projectName)
+					.addFormDataPart(
+							PROJECT_DESCRIPTION_TAG,
+							projectDescription)
+					.addFormDataPart(
+							USER_EMAIL,
+							userEmail)
+					.addFormDataPart(
+							PROJECT_CHECKSUM_TAG,
+							md5Checksum)
+					.addFormDataPart(
+							Constants.TOKEN,
+							token)
+					.addFormDataPart(
+							Constants.USERNAME,
+							username)
+					.addFormDataPart(
+							DEVICE_LANGUAGE,
+							language)
+					.build();
```

Comment:
Maybe I'll go ahead and explain a few changes. I killed the `ConnectionWrapper`, because it was another level of indirection that – in my opinion – did not add any value. Also it was hard to see _what_ exactly was being POSTed, because things kept being added to a map before it was actually used.

The way I formatted the code is of course very verbose, but I'd like to think of it as being easily readable. I think it's a lot better than before, but still has room for improvement.